airbyte-integrations/connectors/source-intercom/source_intercom/components.py

```diff
@@ -0,0 +1,341 @@
+            },
+        }
+        self._update_substream_cursor(stream_slice, last_record)
+
+    def read_parent_stream(
+        self, sync_mode: SyncMode, cursor_field: Optional[str], stream_state: Mapping[str, Any]
+    ) -> Iterable[Mapping[str, Any]]:
+
+        for parent_slice in self.parent_stream.stream_slices(sync_mode=sync_mode, cursor_field=cursor_field, stream_state=stream_state):
+            empty_parent_slice = True
+
+            # update slice with parent state, to pass the initial parent state to the parent instance
+            # stream_state is being replaced by empty object, since the parent stream is not directly initiated
+            parent_prior_state = self._cursor.get("prior_state", {}).get(self.parent_stream_name, {}).get(self.parent_cursor_field)
+            parent_slice.update({"prior_state": {self.parent_cursor_field: parent_prior_state}})
+
+            for parent_record in self.parent_stream.read_records(
+                sync_mode=sync_mode, cursor_field=cursor_field, stream_slice=parent_slice, stream_state=stream_state
+            ):
+                # Skip non-records (eg AirbyteLogMessage)
+                if isinstance(parent_record, AirbyteMessage):
+                    if parent_record.type == Type.RECORD:
+                        parent_record = parent_record.record.data
+
+                try:
+                    substream_slice = dpath.util.get(parent_record, self.parent_field)
+                except KeyError:
+                    pass
+                else:
+                    empty_parent_slice = False
+                    slice = {
+                        self.substream_slice_field: substream_slice,
+                        self.cursor_field.eval(self.config): self._cursor.get(self.cursor_field.eval(self.config)),
+                        self.parent_stream_name: {
+                            self.parent_cursor_field: self._cursor.get(self.parent_stream_name, {}).get(self.parent_cursor_field)
+                        },
+                    }
+                    self._update_parent_cursor(slice, parent_record)
+                    yield slice
+
+            # If the parent slice contains no records,
+            if empty_parent_slice:
+                yield from []
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: Mapping[str, Any]) -> Iterable[Mapping[str, Any]]:
+        stream_state = self.initial_state or {}
+        parent_state = stream_state.get(self.parent_stream_name, {})
+        parent_state.update(**{"prior_state": self._cursor.get("prior_state", {}).get(self.parent_stream_name, {})})
+        slices_generator = self.read_parent_stream(self.parent_sync_mode, self.parent_cursor_field, parent_state)
+        if self.parent_complete_fetch:
+            yield from [slice for slice in slices_generator]
+        else:
+            yield from slices_generator
+
+
+@dataclass
+class LoadBasedWaitTimeFromHeaderBackoffStrategy(BackoffStrategy, JsonSchemaMixin):
```

Comment:
is this %-based-backoff (instead of just waiting for 429) to preserve backwards compatibility? Are intercom rate limits account-based?

Reply:
Yes, @sherifnada this  %-based-backoff is a legacy component from `source-intercom` (CDK version) to preserve BC, Intercom Rate Limits are based on the app, associated with the token, basically rate limits are per app or per account, let's say.